libr/bin/format/elf/elf.c

```diff
@@ -1277,454 +1296,499 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
-		int len = r_buf_read_at (bin->b, s->offset + s->size, buf, sizeof (buf));
-		if (len != sizeof (buf)) {
-			// oops
-		}
-		base = r_mem_mem_aligned (buf, sizeof (buf), (const ut8 *)"\x3c\x0f\x00", 3, 4);
-		if (base) {
-			plt_addr += (int)(size_t) (base - buf);
-		} else {
-			plt_addr += RISCV_PLT_OFFSET; // HARDCODED HACK
-		}
-		plt_addr += rel->k * 16;
-		return plt_addr;
-	}
-	if (plt_section) {
-		const int sizeOfProcedureLinkageTable = 32;
-		const int sizeOfPltEntry = 16;
-		return plt_section->rva + sizeOfProcedureLinkageTable + (rel->k * sizeOfPltEntry);
+static RBinElfReloc *populate_relocs_record(ELFOBJ *bin) {
+	size_t i = 0;
+	size_t num_relocs = get_num_relocs_approx (bin);
+	RBinElfReloc *relocs = calloc (num_relocs + 1, sizeof (RBinElfReloc));
+
+	i = populate_relocs_record_from_dynamic (bin, relocs, i);
+	i = populate_relocs_record_from_section (bin, relocs, i);
+	relocs[i].last = 1;
+
+	bin->reloc_num = i;
+
+	return relocs;
+}
+
+RBinElfReloc *Elf_(r_bin_elf_get_relocs) (ELFOBJ *bin) {
+	if (!bin || !bin->g_sections) {
+		return NULL;
 	}
-	eprintf ("Unsupported relocs type %" PFMT64u " for arch %d\n",
-		(ut64)REL_TYPE, bin->ehdr.e_machine);
-	return UT64_MAX;
+
+	RBinElfReloc *ret = populate_relocs_record (bin);
+
+	return ret;
+}
+
+static void rel_cache_free(HtUPKv *kv) {
+	free (kv->value);
 }
 
+static HtUP *rel_cache_new(ELFOBJ *bin) {
+	RBinElfReloc *relocs = Elf_(r_bin_elf_get_relocs) (bin);
+	const int htsize = R_MIN (bin->reloc_num, 1024);
+	HtUP *rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	size_t i;
 
-static ut64 get_import_addr_mips(ELFOBJ *bin, struct ht_rel_t *rel, RBinElfSection *plt_section) {
-	RBinElfSection *s = get_section_by_name (bin, ".rela.plt");
-	if (s) {
-		ut8 buf[1024];
-		const ut8 *base;
-		ut64 plt_addr = s->rva + s->size;
-		int len = r_buf_read_at (bin->b, s->offset + s->size, buf, sizeof (buf));
-		if (len != sizeof (buf)) {
-			// oops
-		}
-		base = r_mem_mem_aligned (buf, sizeof (buf), (const ut8 *)"\x3c\x0f\x00", 3, 4);
-		if (base) {
-			plt_addr += (int)(size_t) (base - buf);
-		} else {
-			plt_addr += MIPS_PLT_OFFSET + 8; // HARDCODED HACK
+	for (i = 0; i < bin->reloc_num; i++) {
+		RBinElfReloc *tmp = R_NEW (RBinElfReloc);
+		memcpy (tmp, relocs + i, sizeof (RBinElfReloc));
+		if (!ht_up_insert (rel_cache, tmp->sym, tmp)) {
+			free (tmp);
 		}
-		plt_addr += rel->k * 16;
-		return plt_addr;
 	}
-	if (plt_section) {
-		const int sizeOfProcedureLinkageTable = 32;
-		const int sizeOfPltEntry = 16;
-		return plt_section->rva + sizeOfProcedureLinkageTable + (rel->k * sizeOfPltEntry);
-	}
-	eprintf ("Unsupported relocs type %" PFMT64u " for arch %d\n",
-		(ut64)REL_TYPE, bin->ehdr.e_machine);
-	return UT64_MAX;
+	free (relocs);
+	return rel_cache;
 }
 
-static ut64 get_import_addr(ELFOBJ *bin, int sym) {
-	int nrel;
-
-	if ((!bin->shdr || !bin->strtab) && !bin->phdr) {
+static ut64 get_got_entrie(ELFOBJ *bin, RBinElfReloc *rel) {
```

Comment:
typo? `get_got_entries`?

Reply:
I get the value of a got entrie, in some arch the addr of the plt is in the got.

Reply:
yep, I'm just saying this should probably be either `get_got_entry` (singular) or `get_got_entries` (plural) ;)